### PR TITLE
feat(tune): Optuna tuning harness with LLM sampler and age fixes

### PR DESCRIPTION
**Pull request by paulnsorensen**
## Summary

- **Tuning harness** (`vaudeville tune <rule>`) — Optuna-based prompt optimization that toggles example subsets to maximize precision/recall on a held-out split. Uses LLMSampler with TPE fallback, NSGA-II for constrained multi-objective search, and optional LLM-driven candidate authoring.
- **Structured examples** in rule YAML — examples and candidates are now first-class fields with `{{ examples }}` template rendering, replacing inline prompt text.
- **Eval enhancements** — `--json` flag for per-case JSONL output, daemon-first backend selection, cross-validation support.
- **Observability** — daemon-backed inference client, event logging, stats aggregation, live watch TUI.
- **Age review fixes** — 17 items from staff-engineer review: wired LLMSampler/NSGA-II/authoring into study loop, split oversized files (eval.py, harness.py), reduced parameter counts via TrialContext dataclass, fixed nesting violations, cleaned encapsulation (barrel file imports), guarded `split_prompt` and `write_candidates` edge cases, de-slopped silent exception swallowing and cargo-cult noqa comments.

## Test plan

- [x] `just check` passes (ruff format + ruff check + mypy strict)
- [x] `just test` passes (774 passed, 5 skipped)
- [ ] Manual: `vaudeville tune hedging-detector --budget 3 --no-daemon`
- [ ] Manual: `vaudeville watch` with daemon running

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: vaudeville/server/daemon_backend.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    data = bytearray()
+    while True:
+        scan_from = len(data)
+        chunk = sock.recv(RECV_CHUNK)
+        if not chunk:
+            break
+        data.extend(chunk)
+        if data.find(b"\n", scan_from) >= 0:
+            break
+    result: dict[str, object] = json.loads(bytes(data).decode().strip())
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply vaudeville/server/daemon_backend.py]: Fixed — daemon_backend._recv_response now parses only up to the first newline, handling buffered or pipelined data correctly. Commit 0788691.

File: vaudeville/core/rules.py [hunk 2]
```diff
@@ -205,6 +169,20 @@ def resolve_context(
...
+            id=str(e["id"]),
+            input=str(e["input"]),
+            label=str(e["label"]),
+            reason=str(e["reason"]),
+        )
+        for e in raw
+        if isinstance(e, dict)
+    ]
+
+
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply vaudeville/core/rules.py]: Fixed — _parse_examples now validates raw is a list and checks each dict for required keys before accessing them, preventing KeyError on malformed YAML. Non-dict entries are still silently skipped (existing behavior). Commit 0788691.

File: vaudeville/tune/pool.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        with open(path) as f:
+            data = yaml.safe_load(f)
+        if data is not None and not isinstance(data, dict):
+            logger.warning(
+                "Corrupt candidates file %s (not a dict), skipping write", path
+            )
+            return
+        if isinstance(data, dict):
+            existing = data.get("candidates", [])
+    except FileNotFoundError:
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply vaudeville/tune/pool.py:159]: Fixed — write_candidates now validates that the candidates field is a list before appending to it, with a warning and reset if corrupted. Commit 0788691.

File: vaudeville/tune/study.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    passed: bool
+    p_tune: float
+    r_tune: float
+    p_held: float
+    r_held: float
+    trials_run: int
+    pool_size: int
+    best_ids: list[str]
+    study_uri: str
+    diff_path: str
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply vaudeville/tune/study.py:50]: False positive — `study_uri: str` is already defined on line 49 of TuneVerdict. Line 50 is `diff_path: str`. The field is present and used correctly in harness.py.

File: vaudeville/tune/harness.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    p_held, r_held, _ = _eval_subset(ctx.rule, selected, ctx.held_cases, ctx.backend)
+
+    trial.set_user_attr("precision_tune", p_tune)
+    trial.set_user_attr("recall_tune", r_tune)
+    trial.set_user_attr("example_ids", selected)
+
+    if p_held < ctx.config.p_min:
+        trial.set_user_attr("constraint_violated", True)
+
+    return r_held, p_held
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply vaudeville/tune/harness.py]: Fixed — `constraint_violated` now flags when either `p_held < p_min` or `r_held < r_min`, giving NSGA-II symmetric enforcement of both thresholds. Commit 0332ede.

File: vaudeville/tune/study.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    try:
+        import anthropic
+
+        client = anthropic.Anthropic()
+        return LLMSampler(anthropic_client=client)
+    except Exception:
+        logger.debug("Anthropic client unavailable, using NSGA-II sampler")
+        return optuna.samplers.NSGAIISampler(
+            constraints_func=_constraints_func,
+        )
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply vaudeville/tune/study.py:88]: Fixed — added `exc_info=True` to the debug log so the actual exception is visible when troubleshooting. Keeping the broad `except Exception` since multiple failure modes are possible (ImportError, auth errors, network issues) and fail-open to NSGA-II is intentional. Commit 0332ede.
